tensorflow/compiler/mlir/hlo/lib/Transforms/copy_removal.cc

```diff
@@ -0,0 +1,209 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "mlir-hlo/Analysis/userange_analysis.h"
+#include "mlir-hlo/Dialect/mhlo/IR/lhlo_ops.h"
+#include "mlir-hlo/Transforms/PassDetail.h"
+#include "mlir-hlo/Transforms/passes.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/BufferUtils.h"
+
+namespace mlir {
+
+namespace {
+
+class CopyRemoval : BufferPlacementTransformationBase {
+public:
+  explicit CopyRemoval(Operation *op)
+      : BufferPlacementTransformationBase(op), userange(op, allocs, aliases),
+        dominators(op) {}
+
+  void removeCopy(Operation *op) {
+    DenseMap<Value, UseInterval::Vector> useIntervalMap;
+    fillProcessSet(useIntervalMap);
+
+    while (!toProcess.empty()) {
+      // Get the first item in the toProcess set and pop it.
+      Value currentValue = toProcess.front();
+      toProcess.remove(currentValue);
+
+      UseInterval::Vector &currentIntervals = useIntervalMap[currentValue];
+      // The currentInterval is a reference of the UseInterval that is updated
+      // throughout the copy removal.
+      for (UseInterval &currentInterval : currentIntervals) {
+        size_t lastUseId = currentInterval.right;
+        Operation *currentLastUse = userange.getOperation(lastUseId);
+
+        // Check if the currentLastUse is already in the remove list.
+        if (toErase.find(currentLastUse) != toErase.end())
+          continue;
+
+        // Check if currentLastUse implements a CopyOpInterface.
+        auto copyOpInterface = dyn_cast<CopyOpInterface>(currentLastUse);
+        if (!copyOpInterface)
+          continue;
+
+        // Check if the source is the currentValue.
+        if (copyOpInterface.getSource() != currentValue)
+          continue;
+
+        Value copyTarget = copyOpInterface.getTarget();
+        UseInterval::Vector &targetIntervals = useIntervalMap[copyTarget];
+        // Find the UseInterval of the target that contains the copy operation.
+        UseInterval *targetInterval =
+            llvm::find_if(targetIntervals, [=](const UseInterval &i) {
+              return i.contains(lastUseId);
+            });
+
+        // Check if all target uses in the interval are domitated by the
+        // current last use.
+        if (!checkDominance(currentLastUse, lastUseId, copyTarget,
```

Comment:
Why only in that interval? I would have thought that the use range of the target and all its aliases may not conflict with the source. Otherwise, this will not work for loops where the buffer would flow back to an earlier use position.